proxy.py

```diff
@@ -2,225 +2,145 @@
-    )
-
-
-@app.route('/ycm/<process_uuid>/errors', methods=['POST'])
-def get_errors(process_uuid):
-    if process_uuid not in mapping:
-        return "Not found", 404
-    ycms = mapping[process_uuid]
-    data = request.get_json(force=True)
-    if 'patches' in data:
-        ycms.filesync.apply_patches(data['patches'])
-    errors = {}
-    for platform, ycm in sorted(ycms.ycms.iteritems(), reverse=True):
-        result = ycm.parse(data['file'], data['line'], data['ch'])
-        if result is None:
-            continue
-        for error in result:
-            error_key = (error['kind'], error['location']['line_num'], error['text'])
-            if error_key in errors:
-                errors[error_key]['platforms'].append(platform)
-            else:
-                error['platforms'] = [platform]
-                errors[error_key] = error
-
-    return jsonify(
-        errors=errors.values()
-    )
-
-
-@app.route('/ycm/<process_uuid>/goto', methods=['POST'])
-def go_to(process_uuid):
-    if process_uuid not in mapping:
-        return "Not found", 404
-    ycms = mapping[process_uuid]
-    data = request.get_json(force=True)
-    if 'patches' in data:
-        ycms.filesync.apply_patches(data['patches'])
-    for platform, ycm in sorted(ycms.ycms.iteritems(), reverse=True):
-        ycm.parse(data['file'], data['line'], data['ch'])
-        result = ycm.go_to(data['file'], data['line'], data['ch'])
-        if result is not None:
-            return jsonify(location=result)
-    return jsonify(location=None)
-
-
-@app.route('/ycm/<process_uuid>/create', methods=['POST'])
-def create_file(process_uuid):
-    if process_uuid not in mapping:
-        return "Not found", 404
-    ycms = mapping[process_uuid]
-    data = request.get_json(force=True)
-    ycms.filesync.create_file(data['filename'], data['content'])
-    return 'ok'
-
-
-@app.route('/ycm/<process_uuid>/delete', methods=['POST'])
-def delete_file(process_uuid):
-    if process_uuid not in mapping:
-        return "Not found", 404
-    ycms = mapping[process_uuid]
-    data = request.get_json(force=True)
-    ycms.filesync.delete_file(data['filename'])
-    return 'ok'
-
-
-@app.route('/ycm/<process_uuid>/ping', methods=['POST'])
-def ping(process_uuid):
-    if process_uuid not in mapping:
-        return "Not found", 404
-    for ycm in mapping[process_uuid].ycms.itervalues():
-        if not ycm.ping():
-            return 'failed', 504
-
-    return 'ok'
 
+                send(id, result)
+        except (websocket.WebSocketException, geventwebsocket.WebSocketError, TypeError) as e:
+            # WebSocket closed
+            alive[0] = False
+        except Exception as e:
+            alive[0] = False
+            raise
+
+    # Functions to send back a response to a message, with its message ID.
+    def send_response(id, response):
+        if not isinstance(response, collections.Mapping):
+            response = dict(message=response)
```

Comment:
The methods sometimes return a tuple containing a response and an HTTP status code. Is this the correct way to handle that?